### PR TITLE
feat: limit number of light nodes

### PR DESCRIPTION
**Pull request by acud**
This PR limits the number of lightnodes we allow to connect inbound. This is to give some safety level to full nodes operators not to be resource overwhelmed by too many light node requests. This sets the initial limit to 100 above which the node will start kicking away other light nodes to accommodate for the new ones.

---
This change is [Reviewable](https://reviewable.io/reviews/ethersphere/bee/1898)

### CODE REVIEW

File: pkg/p2p/libp2p/libp2p.go [hunk 1]
```diff
@@ -238,141 +244,161 @@ func New(ctx context.Context, signer beecrypto.Signer, networkID uint64, overlay
...
-					s.protocolsmu.RUnlock()
+			if s.lightNodes.Count() > s.lightNodeLimit {
+				// kick another node to fit this one in
+				p, err := s.lightNodes.RandomPeer(peer.Address)
+				if err != nil {
+					s.logger.Debugf("stream handler: cant find a peer slot for light node: %v", err)
+					_ = s.Disconnect(peer.Address)
+					return
+				} else {
+					_ = s.Disconnect(p)
```
Comment (aloknerurkar) [inline pkg/p2p/libp2p/libp2p.go:373]: Can we add a log here?

File: pkg/topology/lightnode/container_test.go [hunk 2]
```diff
@@ -33,14 +33,35 @@ func TestContainer(t *testing.T) {
...
+		}
+		p, err := c.RandomPeer(p1)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if !p.Equal(p2) {
+			t.Fatalf("expected p1 but got %s", p.String())
+		}
+
+		p, err = c.RandomPeer(swarm.NewAddress([]byte("456")))
```
Comment (aloknerurkar) [inline pkg/topology/lightnode/container_test.go:58]: Better to use the variable 'p2' here for better understandibility

File: pkg/topology/lightnode/container.go [hunk 3]
```diff
@@ -49,6 +51,53 @@ func (c *Container) Disconnected(peer p2p.Peer) {
...
 }
 
+func (c *Container) Count() int {
+	c.peerMu.Lock()
+	defer c.peerMu.Unlock()
+
+	return c.count()
+}
+func (c *Container) count() (count int) {
+	_ = c.connectedPeers.EachBinRev(func(peer swarm.Address, _ uint8) (bool, bool, error) {
```
Comment (esadakar) [inline pkg/topology/lightnode/container.go:61]: we can also do `connectedPeers.Length()`, right?

File: pkg/topology/lightnode/container.go [hunk 4]
```diff
@@ -49,6 +51,53 @@ func (c *Container) Disconnected(peer p2p.Peer) {
 	}
 }
 
+func (c *Container) Count() int {
+	c.peerMu.Lock()
```
Comment (esadakar) [inline pkg/topology/lightnode/container.go:55]: Is a lock necessary here?

**Review by esadakar — APPROVED** [review]

**Review by mrekucci — COMMENTED** [review]
> Reviewed 6 of 6 files at r1, 2 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1898)* status: 5 of 6 files reviewed, 5 unresolved discussions (waiting on @acud and @mrekucci)

___
*[pkg/p2p/libp2p/libp2p.go, line 51 at r1](https://reviewable.io/reviews/ethersphere/bee/1898#-MamG7wFBtmGSiw7GocO:-MamG7wG51PvuM_ieaS0:bn451o4) ([raw file](https://github.com/ethersphere/bee/blob/c96cc70764362595b28b15d72cbad1ac5a1ad0c1/pkg/p2p/libp2p/libp2p.go#L51)):*
> ```Go
> )
> 
> const defaultLightnodeLimit = 100
> ```

`defaultLightnodeLimit` -> `defaultLightNodeLimit`
___
*[pkg/p2p/libp2p/libp2p.go, line 91 at r1](https://reviewable.io/reviews/ethersphere/bee/1898#-MamGVCaC1tK4M9VJcA8:-MamGVCaC1tK4M9VJcA9:b-lp751g) ([raw file](https://github.com/ethersphere/bee/blob/c96cc70764362595b28b15d72cbad1ac5a1ad0c1/pkg/p2p/libp2p/libp2p.go#L91)):*
> ```Go
> 	Standalone     bool
> 	FullNode       bool
> 	LightnodeLimit int
> ```

`LightnodeLimit` -> `LightNodeLimit`
___
*[pkg/p2p/libp2p/libp2p.go, line 247 at r1](https://reviewable.io/reviews/ethersphere/bee/1898#-MamGpdI-jovV-qJDN-o:-MamGpdI-jovV-qJDN-p:bm35l7k) ([raw file](https://github.com/ethersphere/bee/blob/c96cc70764362595b28b15d72cbad1ac5a1ad0c1/pkg/p2p/libp2p/libp2p.go#L247)):*
> ```Go
> 	peerRegistry.setDisconnecter(s)
> 
> 	if o.LightnodeLimit == 0 {
> ```

Since you don't want to have a negative limit, I'd consider to rewrite this as:

```golang
	s.lightNodeLimit = defaultLightnodeLimit
	if o.LightnodeLimit > 0 {
		s.lightNodeLimit = o.LightnodeLimit
	}
```

File: pkg/p2p/libp2p/libp2p.go (hunk 1 above)
Comment (acud) [reply pkg/p2p/libp2p/libp2p.go:373]: no need to log - this is executed when there's no error

**Review by acud — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1898)* status: 1 of 6 files reviewed, 4 unresolved discussions (waiting on @aloknerurkar, @esadakar, and @mrekucci)

___
*[pkg/p2p/libp2p/libp2p.go, line 247 at r1](https://reviewable.io/reviews/ethersphere/bee/1898#-MamGpdI-jovV-qJDN-o:-Mb-sVD54BeMW5SLeRx7:b-lwyc4w) ([raw file](https://github.com/ethersphere/bee/blob/c96cc70764362595b28b15d72cbad1ac5a1ad0c1/pkg/p2p/libp2p/libp2p.go#L247)):*
<details><summary><i>Previously, mrekucci (Peter Mrekaj) wrote…</i></summary><blockquote>

Since you don't want to have a negative limit, I'd consider to rewrite this as:

```golang
	s.lightNodeLimit = defaultLightnodeLimit
	if o.LightnodeLimit > 0 {
		s.lightNodeLimit = o.LightnodeLimit
	}
```
</blockquote></details>

good suggestion. nice!
___
*[pkg/p2p/libp2p/libp2p.go, line 373 at r1](https://reviewable.io/reviews/ethersphere/bee/1898#-Malwuk_8nvCs6d3jIQi-r1-373:-Mb-sbfS7k_pSWG8qc5-:b-896fix) ([raw file](https://github.com/ethersphere/bee/blob/c96cc70764362595b28b15d72cbad1ac5a1ad0c1/pkg/p2p/libp2p/libp2p.go#L373)):*
<details><summary><i>Previously, aloknerurkar (Alok Nerurkar) wrote…</i></summary><blockquote>

Can we add a log here?
</blockquote></details>

Done.
___
*[pkg/topology/lightnode/container.go, line 55 at r1](https://reviewable.io/reviews/ethersphere/bee/1898#-Malwuk_8nvCs6d3jIQk-r1-55:-Mb-tYUB61zDqWWdq-kO:b-896fix) ([raw file](https://github.com/ethersphere/bee/blob/c96cc70764362595b28b15d72cbad1ac5a1ad0c1/pkg/topology/lightnode/container.go#L55)):*
<details><summary><i>Previously, esadakar (Esad Akar) wrote…</i></summary><blockquote>

Is a lock necessary here?
</blockquote></details>

Done.
___
*[pkg/topology/lightnode/container.go, line 61 at r1](https://reviewable.io/reviews/ethersphere/bee/1898#-Malwuk_8nvCs6d3jIQk-r1-61:-Mb-tY1--fJKl7lW--ah:b-896fix) ([raw file](https://github.com/ethersphere/bee/blob/c96cc70764362595b28b15d72cbad1ac5a1ad0c1/pkg/topology/lightnode/container.go#L61)):*
<details><summary><i>Previously, esadakar (Esad Akar) wrote…</i></summary><blockquote>

we can also do `connectedPeers.Length()`, right?
</blockquote></details>

Done.
___
*[pkg/topology/lightnode/container_test.go, line 58 at r1](https://reviewable.io/reviews/ethersphere/bee/1898#-Malwuk_8nvCs6d3jIQl-r1-58:-Mb-tYxkDd-u46bfd1Za:b-896fix) ([raw file](https://github.com/ethersphere/bee/blob/c96cc70764362595b28b15d72cbad1ac5a1ad0c1/pkg/topology/lightnode/container_test.go#L58)):*
<details><summary><i>Previously, aloknerurkar (Alok Nerurkar) wrote…</i></summary><blockquote>

Better to use the variable 'p2' here for better understandibility
</blockquote></details>

Done.

**Review by zelig — COMMENTED** [review]
> Reviewed 1 of 6 files at r1, 2 of 3 files at r2, 2 of 4 files at r3.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1898)* status: 4 of 6 files reviewed, 5 unresolved discussions (waiting on @acud, @aloknerurkar, @esadakar, and @mrekucci)

___
*[pkg/topology/lightnode/container.go, line 58 at r3](https://reviewable.io/reviews/ethersphere/bee/1898#-Mb0-t7y96s71C65hxGh:-Mb0-t7y96s71C65hxGi:bcyzopx) ([raw file](https://github.com/ethersphere/bee/blob/4914f75a449907b08df56103538b75c4aa36c0aa/pkg/topology/lightnode/container.go#L58)):*
> ```Go
> }
> 
> func (c *Container) RandomPeer(not swarm.Address) (swarm.Address, error) {
> ```

can you explain why this is not just:
```
i := rand.Intn(c.connectedPeers.Length())
_ = c.connectedPeers.EachBinRev(func(peer swarm.Address, _ uint8) (bool, bool, error) {
	if 0 == i {
		addr = peer
		return true, false, nil
	}
	i++
	return false, false, nil
})

```

**Review by zelig — COMMENTED** [review]
> Reviewed 2 of 4 files at r3.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1898)* status: all files reviewed, 5 unresolved discussions (waiting on @acud, @aloknerurkar, and @esadakar)

**Review by acud — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1898)* status: all files reviewed, 5 unresolved discussions (waiting on @aloknerurkar, @esadakar, and @zelig)

___
*[pkg/topology/lightnode/container.go, line 58 at r3](https://reviewable.io/reviews/ethersphere/bee/1898#-Mb0-t7y96s71C65hxGh:-Mb0SGrS76ByNLHMUlJn:b75k7vy) ([raw file](https://github.com/ethersphere/bee/blob/4914f75a449907b08df56103538b75c4aa36c0aa/pkg/topology/lightnode/container.go#L58)):*
<details><summary><i>Previously, zelig (Viktor Trón) wrote…</i></summary><blockquote>

can you explain why this is not just:
```
i := rand.Intn(c.connectedPeers.Length())
_ = c.connectedPeers.EachBinRev(func(peer swarm.Address, _ uint8) (bool, bool, error) {
	if 0 == i {
		addr = peer
		return true, false, nil
	}
	i++
	return false, false, nil
})

```
</blockquote></details>

because `crypto/rand`

**Review by esadakar — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1898)* status: all files reviewed, 5 unresolved discussions (waiting on @aloknerurkar, @esadakar, and @zelig)

**Review by aloknerurkar — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1898#-:-Mb1Sdaf4BBO3pPj-45Q:bnfp4nl)* status: all files reviewed, 5 unresolved discussions (waiting on @aloknerurkar, @esadakar, and @zelig)

**Pull request merged**